### ISSUE TITLE
Using incomplete types in Rust

### ISSUE THREAD

**Issue by Ayush1325** [r-efi/r-efi#41]
I am trying to use the [`r_efi::protocols::file::Info`](https://github.com/r-efi/r-efi/blob/34d3da6498bdec2c0fb1a6e1b2f00c8affbcbeed/src/protocols/file.rs#L58), in the [`r_efi::protocols::file::Protocol::get_info`](https://github.com/r-efi/r-efi/blob/34d3da6498bdec2c0fb1a6e1b2f00c8affbcbeed/src/protocols/file.rs#L176) function. However, I don't really understand how to create this type on the Rust side. 

So are there any examples of how to use ZST in ffi?

**Comment by dvdhrm** [comment]
`File::get_info` was one of those that requires the caller to preallocate, wasn't it? In that case, you have to create a buffer big enough to hold `File::Info` plus a variable amount of bytes for the file-name (e.g., 1024). You then pass that buffer to `get_info()`. If the function returns with a size-error, you increase the buffer-size and try again (this time allocating on the heap).

Once the function returns with success, you transmute the buffer to `File::Info` (make sure the alignment of the buffer is correct, which is always the case for heap allocations). To access the file-name, you calculate the number of trailing bytes by subtracting the object size of the returned length. Then use `std::slice::from_raw_parts()` to create a temporary slice for the file-name.

I don't think there is any example, yet. If you feel like it, add it to ./examples/!

**Comment by Ayush1325** [comment]
Currently, I am doing something like the following:
```rust
let protocol = self.inner.as_ptr();
let mut buf_size = 0usize;
                 
// inner just called the get_info function                                                                    
match inner(protocol, &mut buf_size, crate::ptr::null_mut()) {
    Ok(()) => unreachable!(),
    Err(e) => match e.kind() {
        // Returning this error if EFI_BUFFER_TOO_SMALL status is returned
        io::ErrorKind::InvalidData => {}
        _ => return Err(e),
    },
}
                                                                                     
let layout = Layout::from_size_align(buf_size, 8usize)
    .map_err(|_| io::Error::new(io::ErrorKind::Other, "Invalid buffer size"))?;
let buf: NonNull<file::Info> = Global
    .allocate(layout)
    .map_err(|_| io::Error::new(io::ErrorKind::Other, "Failed to allocate Buffer"))?
    .cast();
                                                                                     
match inner(protocol, &mut buf_size, buf.as_ptr().cast()) {
    Ok(()) => {}
    Err(e) => return Err(e),
}
                                                                                     
let res = FileAttr::from(unsafe { buf.as_ref() });
                                                                                     
unsafe {
    Global.deallocate(buf.cast(), layout);
}
```

Also, I was thinking wouldn't it be better to use const generic for such structures? Something like:
```rust
#[repr(C)]
#[derive(Clone, Copy, Debug)]
pub struct Info<const N: usize> {
    pub size: u64,
    pub file_size: u64,
    pub physical_size: u64,
    pub create_time: crate::system::Time,
    pub last_access_time: crate::system::Time,
    pub modification_time: crate::system::Time,
    pub attribute: u64,
    pub file_name: [crate::base::Char16; N],
}
```

This should make it possible to have stack-allocated Info<256> and Heap allocated Info<0>. Or are there any specific reasons why this is not possible/should not be done?

**Comment by dvdhrm** [comment]
> ```rust
> #[repr(C)]
> #[derive(Clone, Copy, Debug)]
> pub struct Info<const N: usize> {
>     pub size: u64,
>     pub file_size: u64,
>     pub physical_size: u64,
>     pub create_time: crate::system::Time,
>     pub last_access_time: crate::system::Time,
>     pub modification_time: crate::system::Time,
>     pub attribute: u64,
>     pub file_name: [crate::base::Char16; N],
> }
> ```
> 
> This should make it possible to have stack-allocated Info<256> and Heap allocated Info<0>. Or are there any specific reasons why this is not possible/should not be done?

The Const-generic-MVP was not a thing back when we started r-efi. Note that r-efi can be compiled on rust stable and does not require nightly features, so we never considered const generics.

Anyway, I do admit this looks quite neat. Sadly, it does not work with runtime-variables, so we still need to find a way to deal with that. We have not really found a nice way to handle all required cases, so we just carry the most basic static structures so far.

If you feel like this is a nice addition, feel free to open a PR. But note that we need to verify that this is stable rust ABI (I never checked whether the compiler has any guarantees on generic-ABI), and this is an API break, so we would have to bump the major version (if you add a default of 0, this might actually not be an API break. I am unsure.).

Ideally, we just play around with possible implementations for a while and see what feels best. Especially for dynamic lengths at runtime I want to have a neat solution that avoids having to deal with custom allocations. Note that there are some pending rust RFCs dealing with similar issues regarding FFI, but none looked promising making it to stable so far. I also stopped tracking a year ago, so maybe things changed?

Anyway, long story short: I am open to improving the status quo, but r-efi needs to be ABI compatible to the spec, which limits us quite a bit.

**Comment by Ayush1325** [comment]
Sounds reasonable. I will try out a few different solutions for improving the Rust-UEFI interactions once I am finished with allowing Rust CI to run tests for UEFI>.

Will open a PR after I have tested out what works well.

**Comment by dvdhrm** [comment]
The const-generics have now been merged. It is still rather cumbersome to deal with those types with runtime-dynamic sizes, but I assume there is not much we can do about it. Rust-types have to be known at compile-time. I assume a set of helpers is what to go for.

I will close this for now. If more help is required, we can always open a new report!

Thanks for tackling this!
